tapdb/sqlc/queries/assets.sql

```diff
@@ -776,6 +776,86 @@ SELECT key_family, key_index
 FROM internal_keys
 WHERE raw_key = $1;
 
+-- name: UpsertTapscriptTreeRootHash :one
+INSERT INTO tapscript_roots (
+    root_hash, branch_only
+) VALUES (
+    $1, $2
+) ON CONFLICT (root_hash)
+    -- This is a NOP, the root_hash is the unique field that caused the
+    -- conflict. The tree should be deleted before switching between branch and
+    -- leaf storage.
+    DO UPDATE SET root_hash = EXCLUDED.root_hash
+RETURNING root_id;
+
+-- name: UpsertTapscriptTreeNode :one
+INSERT INTO tapscript_nodes (
+    raw_node
+) VALUES (
+    $1
+) ON CONFLICT (raw_node)
+    -- This is a NOP, raw_node is the unique field that caused the conflict.
+    DO UPDATE SET raw_node = EXCLUDED.raw_node
+RETURNING node_id;
+
+-- name: UpsertTapscriptTreeEdge :one
+INSERT INTO tapscript_edges (
+    root_hash_id, node_index, raw_node_id
+) VALUES (
+    $1, $2, $3
+) ON CONFLICT (root_hash_id, node_index, raw_node_id)
+    -- This is a NOP, root_hash_id, node_index, and raw_node_id are the unique
+    -- fields that caused the conflict.
+    DO UPDATE SET root_hash_id = EXCLUDED.root_hash_id,
+    node_index = EXCLUDED.node_index, raw_node_id = EXCLUDED.raw_node_id
+RETURNING edge_id;
+
+-- name: FetchTapscriptTreeRoots :many
+SELECT *
+FROM tapscript_roots;
+
+-- name: FetchTapscriptTree :many
+WITH tree_info AS (
+    -- We use this CTE to fetch all edges that link the given tapscript tree
+    -- root hash to child nodes. Each edge also contains the index of the child
+    -- node in the tapscript tree.
+    SELECT tapscript_roots.root_id, tapscript_roots.root_hash,
+        tapscript_roots.branch_only, tapscript_edges.node_index,
+        tapscript_edges.raw_node_id
+    FROM tapscript_roots
+    JOIN tapscript_edges
+        ON tapscript_roots.root_id = tapscript_edges.root_hash_id
+    WHERE tapscript_roots.root_hash = @root_hash
+)
+-- Deocoding of the tree nodes differs if they are TapLeaf of TapHash objects.
+-- We also sort by node_index here vs. having the caller sort the query results.
+SELECT tree_info.branch_only, tapscript_nodes.raw_node
+FROM tapscript_nodes
+JOIN tree_info
+    ON tree_info.raw_node_id = tapscript_nodes.node_id
+ORDER BY tree_info.node_index ASC;
+
+-- name: DeleteTapscriptTreeEdges :exec
+WITH tree_info AS (
+    -- We use this CTE to fetch all edges that link the given tapscript tree
+    -- root hash to child nodes.
+    SELECT tapscript_roots.root_id, tapscript_roots.root_hash,
+        tapscript_edges.edge_id
+    FROM tapscript_roots
+    JOIN tapscript_edges
+        ON tapscript_roots.root_id = tapscript_edges.root_hash_id
+    WHERE tapscript_roots.root_hash = @root_hash
+)
+-- TODO(jhb): This query does not delete any nodes. Separate query to delete
+-- nodes with no edges? (Which may be different from all child nodes of this
```

Comment:
As discussed offline: Could use the `WHERE NOT EXISTS(...)` SQL feature for this.